tutorials/llm/llama/domain-adaptive-pretraining/code/tokenization_helper.py

```diff
@@ -0,0 +1,400 @@
+    f = open(domain_tok_vocab_path)
+    vocab = json.load(f)
+    print("Domain vocab size:", len(vocab))
+
+    tokens = []
+    drop_tokens = []
+    print("token pattern: ", token_pattern)
+    for v in vocab:
+        if re.search(token_pattern, v):
+            tokens.append(v.replace("Ġ", "▁"))
+        else:
+            drop_tokens.append(v)
+    print("Num of added tokens and dropped tokens", len(tokens), len(drop_tokens))
+
+    m = model.ModelProto()
+    m.ParseFromString(open(original_tokenizer_path, 'rb').read())
+    print(f'Original model pieces: {len(m.pieces)}')
+    print(m.trainer_spec)
+    # exit()
+    ori_vol = []
+    for piece in m.pieces:
+        ori_vol.append(piece.piece)
+    print("original vocab size: ", len(ori_vol))
+    ori_vol = set(ori_vol)
+    data = set(data)
+
+    new_tokens = []
+    for token in tokens:
+        ## for quick test
+        # if token == tokens[1000]:
+        #    break
+        if token not in ori_vol:
+            token1 = token.replace("▁", " ")
+            occur_cnt = 0
+            flag = True
+            for s in data:
+                if token1 in s:
+                    occur_cnt += 1
+                    if occur_cnt > occur_limit:
+                        flag = False
+                        break
+            if flag:
+                new_tokens.append(token)
+    print("new token cnt: ", len(new_tokens))
+
+    normal_cnt = len(new_tokens)
+    dummy_cnt = (len(new_tokens) // 1024 + 1) * 1024 - len(new_tokens)
+    add_cnt = normal_cnt + dummy_cnt
+    print("add token cnt: ", add_cnt)
+    print("add normal token cnt: ", normal_cnt)
+    print("add dummy token cnt: ", dummy_cnt)
+    assert dummy_cnt >= 3, "should be at least 3 extra tokens for finetuning"
+
+    dummy_tokens = []
+    for i in range(dummy_cnt):
+        dummy_tokens.append(f"<extra_id_{i}>")
+
+    record = []
+    N = len(m.pieces)
+    for i, sym in enumerate(new_tokens):
+        new_sym = m.SentencePiece()
+        new_sym.piece = sym
+        new_sym.score = 0.0  # default score for USER_DEFINED
+        new_sym.type = 4  # type value for USER_DEFINED
+        m.pieces.insert(N + i, new_sym)  # position after default control symbols ("<unk>", "<s>", "</s>")
+        record.append([sym, N + i])
+
+    N = len(m.pieces)
+    for i, sym in enumerate(dummy_tokens):
+        new_sym = m.SentencePiece()
+        new_sym.piece = sym
+        new_sym.score = 0.0  # default score for USER_DEFINED
+        new_sym.type = 4  # type value for USER_DEFINED
+        m.pieces.insert(N + i, new_sym)  # position after default control symbols ("<unk>", "<s>", "</s>")
+        record.append([sym, N + i])
+
+    print(f'New model pieces: {len(m.pieces)}')
+    print(m.trainer_spec)
+
+    with open(new_vocab_path, "w", encoding="utf8") as fp:
+        json.dump(record, fp)
+
+    with open(new_model_path, 'wb') as f:
+        f.write(m.SerializeToString())
+
+    if split > 1:
```

Comment:
The rest of function from here on will expand the embedding&output layer by append to original embedding/output, and then save to disk. However: 1. the tokens we obtain here are not filtered/sorted by frequency yet. 2. there is similar function in extend_tokenizer_utils.py:extend_tokenizer_high_freq_tokens that will save the extended embedding for high freq tokens again and write to the same file. Should this part be deleted?

Reply:
@jvamaraju can you pls address this?